File: java/swapnodesinpairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ListNode pPre = new ListNode(0);
        pPre.next = head;
        ListNode pfirst = head;
        ListNode p = head;
        ListNode pnext = head;
        ListNode lastpnext = pPre;
        int no = 0;
        while(p!=null){
            no++;
            if(no==2){
                no=0;
                pnext = p.next;
                pfirst = lastpnext.next;
                lastpnext.next = p;
                p.next = pfirst;
                pfirst.next = pnext;
                lastpnext = pfirst;
                p = pfirst;
            }
            p = p.next;
        }
        return pPre.next;
    }
}

File: cpp/SwapNodesinPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* p = dummy;
        while(true) {
            if(p->next && p->next->next) {
                ListNode* n0 = p->next;
                ListNode* n1 = p->next->next;
                ListNode* n2 = p->next->next->next;
                p->next = n1;
                n1->next = n0;
                n0->next = n2;
                p = n0;
            } else{
                break;
            }
        }
        ListNode* res = dummy->next;
        delete dummy;
        return res;
    }
};
